### ISSUE TITLE
t.conch.insults.window.ScrolledArea is not instantiable.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4197]
|<img alt="ali's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ali reported|
|-|-|
|Trac ID|trac#4197|
|Type|defect|
|Created|2010-01-04 12:35:58Z|
|Branch|https://github.com/twisted/twisted/tree/scrolledarea-init-4197|

Passes something up the inheritance chain wrongly. Patch attached.

NB: The tests for this module are basically non-existent.

Attachments:

* [fix-scrolledarea.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/443/4431bf7894e5c68aab8659fe44a532f1308e8236/b477750e34886d84e6ee647d3b6353dc04df10e7.diff) (1229 bytes) - added by ali on 2010-01-04 12:36:14Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4197 4197
type__defect defect
reporter__ali ali
priority__normal normal
milestone__ 
branch__branches_scrolledarea_init_4197 branches/scrolledarea-init-4197
branch_author__ali__exarkun ali, exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__insults insults
time__1262608558000000 1262608558000000
changetime__1305469512000000 1305469512000000
version__None None
owner__exarkun exarkun
cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to ali|
|-|-|

Hi ali,

Thanks a lot for spotting this.

  1. The test method should be named `test_canInstantiate`: the first letter after the method prefix should be lower-case.
  2. Testing for instantiation isn't really useful by itself.  When first written, tests should _fail_, not _break_; the difference is that you should have an assertion whose prediction is false, not a random exception that gets raised.  The test should assert something about the object resulting from the instantiation.
  3. The test should have a docstring that explains what it's asserting.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

[#4615](https://github.com/twisted/twisted/issues/4615) was a duplicate of this.  There's also some files attached there, but I don't think they're closer to something applicable than what's here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [31798]) Branching to 'scrolledarea-init-4197'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [31799]) Apply ali's patch plus my changes to address review feedback

refs [#4197](https://github.com/twisted/twisted/issues/4197)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

See [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/scrolledarea-init-4197), I expect they'll be good.

As an aside, though it's not strictly the reviewer's responsibility, I think it might have been helpful to recommend what the unit test should do, rather than only pointing out that what it did was not very good.  I had to think for a minute before finding something halfway decent to do aside from instantiate the class, so someone less familiar with the code (ie, someone who didn't write all this code in the first place ;) might have gotten stuck at that point.  I don't know if that's what happened to ali, though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

That was definitely not the best recommendation; unfortunately I didn't know enough about `ScrolledArea` to make a sensible suggestion either.  I was hoping that ali knew better than I did :).  I tried to describe the more desirable "failure, not error" pattern though (and I really would have accepted just about any test that would fail under some condition).

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|ivank set owner to @exarkun|
|-|-|

1. In window.py, this line exceeds 79 columns (per coding-standard.xhtml):

`    A L{ScrolledArea} contains another widget wrapped in a viewport and vertical`

2. In test_window.py, I see "a widget which creates a viewport contained another widget".  Did you mean "contains"?

Otherwise, looks good for merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [31872]) Merge scrolledarea-init-4197

Author: ali, exarkun
Reviewer: glyph, ivank
Fixes: [#4197](https://github.com/twisted/twisted/issues/4197)

Change the `twisted.conch.insults.window.ScrolledArea` initializer to not
pass the contained widget to the parent `__init__`, since the parent does
not expect it and on some versions of Python this results in a TypeError.
